### DISCUSSION TITLE
connectOrCreate is not atomic

### DISCUSSION THREAD

**Discussion by ddavid67** [prisma/prisma#6004] [Q&A]
Hello

I just discovered an issue when seeding my database with lot of random values

I have a classic many-to-many relation with post <-> tag (with explicit tables)
I am using connectOrCreate to create new posts and associate related tags.

If I am creating 2 posts at the same time with the same tag, and this tag doesn't exist yet, one of the creation fail because both creations tries to create the same tag and obviously one of them will create it first and the second will fail because of duplicate unique field (the tagname)

I fixed this by running a createMany with skipDuplicate on the tag table, to make sure the tag already exist before creating the post. 

Is there a better solution ?

Maybe a "retries" option can help with this ?
Because the just running the creation again will be successfull.

**Comment by ryands17** [comment]
@daeschlerdavid 👋 

In this case, I would suggest using `prisma.$transaction` to and all your create posts in those so that it remains atomic. So something like this:

```ts
await prisma.$transaction([
    prisma.post.create({
      data: {
        name: `post 1`,
        tags: {
          connectOrCreate: {
            create: { name: 'tag 1' },
            where: { name: 'tag 1' },
          },
        },
      },
    }),
    prisma.post.create({
      data: {
        name: `post 1`,
        tags: {
          connectOrCreate: {
            create: { name: 'tag 1' },
            where: { name: 'tag 1' },
          },
        },
      },
    }),
])
```

Each call will add it's own transaction block which is why you're facing this issue.

**Comment by ddavid67** [comment]
this will only work if I have full control over concurrency but this not the case in production

the problem remains of the two clients are on separate thread/node process/ or even two separate db connections from the same pool

**Reply by ryands17** [reply]
I tried the following:

```ts
await Promise.all([
    prisma.post.create({
      data: {
        name: `post 2`,
        tags: {
          connectOrCreate: {
            create: { name: 'tag2' },
            where: { name: 'tag2' },
          },
        },
      },
    }),
    prisma.post.create({
      data: {
        name: `post 3`,
        tags: {
          connectOrCreate: {
            create: { name: 'tag2' },
            where: { name: 'tag2' },
          },
        },
      },
    }),
  ])
```

This is inserting in parallel and it works fine.

**Comment by ddavid67** [comment]
I was able to reproduce with this minimal config that is similar to my production :

`schema.prisma` :

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = []
}

model Tag2 {
  id        String       @id @default(cuid())
  name      String       @unique
  createdAt DateTime     @default(now())
  posts     TagsOnPost[]
}

model TagsOnPost {
  post      Post     @relation(fields: [postId], references: [id])
  postId    String
  tag       Tag2     @relation(fields: [tagId], references: [id])
  tagId     String
  createdAt DateTime @default(now())

  @@id([postId, tagId])
}

model Post {
  id   String       @id @default(cuid())
  name String
  tags TagsOnPost[]
}
```

`seed.ts` :

```ts
import { PrismaClient } from '@prisma/client'

const main = async () => {
  const prisma = new PrismaClient()

  const createPostsPromises = Array.from({ length: 2 }).map(() => {
    return prisma.post.create({
      data: {
        name: 'post',
        tags: {
          create: [
            {
              tag: {
                connectOrCreate: {
                  create: { name: 'tag1' },
                  where: { name: 'tag1' }
                }
              }
            }
          ]
        }
      }
    })
  })

  try {
    await Promise.all(createPostsPromises)
  } catch (e) {
    console.log(e)
  } finally {
    await prisma.$disconnect()
  }
}

main()
```

**Result :**

```
david@nr200p ~/d/app (main)> npx prisma db seed --preview-feature
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Running ts-node "prisma/seed.ts" ...
PrismaClientKnownRequestError2 [PrismaClientKnownRequestError]: Unique constraint failed on the fields: (`name`)
    at cb (/home/david/dev/app/node_modules/@prisma/client/runtime/index.js:79160:17)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async Promise.all (index 1)
    at main (/home/david/dev/app/prisma/seed.ts:27:5) {
  code: 'P2002',
  clientVersion: '2.18.0',
  meta: { target: [ 'name' ] }
}

🌱  Your database has been seeded.

```

**Reply by ryands17** [reply]
@daeschlerdavid Tried the same code above on version `2.18.0` and it works fine.

![image](https://user-images.githubusercontent.com/19697099/110444030-94a90080-80e2-11eb-9347-e37e5d19fbe8.png)

I have used the same above schema and logic.

**Reply by ddavid67** [reply]
Your output looks very sequential
The first transaction is commited before the second one tries to look for tag existence

Mine looks like this.
As you see the the first transaction is not commited when the second one is trying to detect tag existence.

```
david@nr200p ~/d/app (main)> npx prisma db seed --preview-feature
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Running ts-node "prisma/seed.ts" ...
prisma:query BEGIN
prisma:query INSERT INTO "public"."Post" ("id","name") VALUES ($1,$2) RETURNING "public"."Post"."id"
prisma:query SELECT "public"."Tag2"."id" FROM "public"."Tag2" WHERE "public"."Tag2"."name" = $1 OFFSET $2
prisma:query INSERT INTO "public"."Tag2" ("id","name","createdAt") VALUES ($1,$2,$3) RETURNING "public"."Tag2"."id"
prisma:query BEGIN
prisma:query INSERT INTO "public"."TagsOnPost" ("postId","tagId","createdAt") VALUES ($1,$2,$3) RETURNING "public"."TagsOnPost"."postId", "public"."TagsOnPost"."tagId"
prisma:query INSERT INTO "public"."Post" ("id","name") VALUES ($1,$2) RETURNING "public"."Post"."id"
prisma:query SELECT "public"."Post"."id", "public"."Post"."name" FROM "public"."Post" WHERE "public"."Post"."id" = $1 LIMIT $2 OFFSET $3
prisma:query SELECT "public"."Tag2"."id" FROM "public"."Tag2" WHERE "public"."Tag2"."name" = $1 OFFSET $2
prisma:query COMMIT
prisma:query INSERT INTO "public"."Tag2" ("id","name","createdAt") VALUES ($1,$2,$3) RETURNING "public"."Tag2"."id"
prisma:query ROLLBACK
PrismaClientKnownRequestError2 [PrismaClientKnownRequestError]: Unique constraint failed on the fields: (`name`)
    at cb (/home/david/dev/app/node_modules/@prisma/client/runtime/index.js:79160:17)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async Promise.all (index 0)
    at main (/home/david/dev/app/prisma/seed.ts:27:5) {
  code: 'P2002',
  clientVersion: '2.18.0',
  meta: { target: [ 'name' ] }
}

🌱  Your database has been seeded
```


How can we explain such difference ?
Also there are little syntax differences between our two logs.  Our environnement doesn't look exactly the same.

**Reply by ryands17** [reply]
I'm on Mac using Prisma version `2.18.0` and on Postgres 12.

> The first transaction is committed before the second one tries to look for tag existence

I'm not sure why that is. Might be a database specific thing.

![image](https://user-images.githubusercontent.com/19697099/110461979-240bdf00-80f6-11eb-81f0-48b4795cb532.png)

**Reply by ddavid67** [reply]
can you try increasing the number of post insert ? (with the length parameter Array.from({length:100})
maybe that would increase the chance of you reproducing the issue

anyway its not a prisma issue. but it would be cool if you could reproduce it 

even if you could reproduce it, i dont think there a solution for this?
to my knowledge there is no such thing as "SELECT for INSERT" to lock something that does not already exist...

**Reply by ryands17** [reply]
`Array.from({length:100})` has the same behaviour unfortunately. I think you're facing a similar issue that's pointed here: https://github.com/prisma/prisma/issues/3242

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
